File: src/structure/package/applyPackageNameMapper.ts
import path from "node:path";
import {Configuration} from "../../configuration/configuration.js";
import {applyMapper} from "../../utils/fileName.js";
import {moduleNameToPackage} from "../module/moduleNameToPackage.js";
import {packageToFileName} from "./packageToFileName.js";

export interface PackageMappingResult {
    fileName: string
    package: string[],
}

export function applyPackageNameMapper(
    packageChunks: string[],
    fileName: string,
    configuration: Configuration,
): PackageMappingResult {
    const {packageNameMapper} = configuration

    const filePath = packageToFileName(packageChunks, fileName)

    const mappedFilePath = applyMapper(filePath, packageNameMapper)

    const dirName = path.dirname(mappedFilePath)
    const baseName = path.basename(mappedFilePath)

    return {
        fileName: baseName,
        package: moduleNameToPackage(dirName)
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/package/applyPackageNameMapper.kt
package io.github.sgrishchenko.karakum.structure.`package`

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.util.applyMapper
import js.array.ReadonlyArray
import kotlinx.js.JsPlainObject
import node.path.path

@JsPlainObject
external interface PackageMappingResult {
    val fileName: String
    val `package`: ReadonlyArray<String>
}

fun applyPackageNameMapper(
    packageChunks: ReadonlyArray<String>,
    fileName: String,
    configuration: Configuration,
): PackageMappingResult {
    val packageNameMapper = configuration.packageNameMapper

    val filePath = packageToFileName(packageChunks, fileName)

    val mappedFilePath = applyMapper(filePath, packageNameMapper)

    val dirName = path.dirname(mappedFilePath)
    val baseName = path.basename(mappedFilePath)

    return PackageMappingResult(
        fileName = baseName,
        `package` = dirNameToPackage(dirName)
    )
}
